Forms: save submissions toggle

vianasw opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes FORMS-247

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* I've added a toggle in the form's block panel to allow users to not save form submissions in wp-admin. The toggle is on by default.
<img width="626" height="556" alt="Screenshot 2025-09-04 at 16 53 47" src="https://github.com/user-attachments/assets/9198599a-bbcc-4860-95c8-9a40cabd8c70" />

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add a form to a post or page.
* Go to the forms block settings' `Manage responses` panel and set the `Save response` toggle off.
* Publish the post/page and then submit a form response.
* The post-submission page will show the form submission data. Also, check that after a page refresh, the form submission is still present.
<img width="1316" height="934" alt="Screenshot 2025-09-08 at 11 54 28" src="https://github.com/user-attachments/assets/2664ea49-ed3d-4d7b-88ff-2ad63b1cef64" />
* Go to the Jetpack Forms responses page in wp-admin and check that the response you submitted doesn't show up there.
* Check that you've received an email with the form submission data. The email shouldn't have the `View in dashboard` button and the `Mark as spam` link.
<img width="1362" height="1124" alt="Screenshot 2025-09-08 at 11 50 07" src="https://github.com/user-attachments/assets/343e9b4e-cd5c-42b0-8375-fe262d824ded" />
* Go back to edit the post/page you've created and now set the `Save responses` toggle on.
* Check that now the responses show up in the Jetpack forms responses page in wp-admin.
* Check that on the post-submission page, the form submission data appears after page refresh.
* Check that the form submission email shows the `View in Dashboard` button and `Mark as spam` link.
<img width="1342" height="1286" alt="Screenshot 2025-09-08 at 11 50 15" src="https://github.com/user-attachments/assets/72edb738-1f86-4910-b3a7-d931c404d833" />
* Set up a form integration, such as MailPoet, and check that the integration continues to work with the `Save response` toggle on and off.

vianasw committed:
Add 'Save response' form toggle

I've added a toggle in the form's block panel to allow users to not save
form submissions in wp-admin. The toggle is on by default.

vianasw committed:
Make saveResponses a yes/no form attribute

Due to shortcode bacwkard render compability, we need to store the
saveResponses attribute as a 'yes'/'no' string attribute.

vianasw committed:
changelog

CGastrell reviewed (changes requested):
While the main goal of the PR works fine, it carries a side effect: the confirmation message is not showing any of the submitted data:

Before:
<img width="750" height="552" alt="image" src="https://github.com/user-attachments/assets/b8c42dc6-d22b-4209-a1e2-6159f677db0e" />

After:
<img width="690" height="324" alt="image" src="https://github.com/user-attachments/assets/a2377446-99b4-4edd-8d4c-38fd1c560b2f" />

CGastrell commented:
I think the same happens to emails:

<img width="676" height="392" alt="image" src="https://github.com/user-attachments/assets/5552b0d1-8cb3-4a73-b361-7724137933fc" />

<img width="712" height="267" alt="image" src="https://github.com/user-attachments/assets/76759bc0-8f8e-4f2e-b6fd-ecdbc859c21b" />

vianasw committed:
Fix forms handling when responses aren't saved

This commit addresses issues in the contact form email functionality
when the "Save responses" toggle is disabled. The changes ensure that:

1. Method signature enhancement: Added an optional $response parameter
   to get_compiled_form_for_email() to allow passing response data
   directly instead of always fetching from database
2. Conditional response fetching: Modified the method to only fetch
   response from database when not provided as parameter, preventing
   errors when no feedback ID exists
3. Defensive URL building: Added checks before building dashboard URLs
   and spam marking URLs to handle cases where $post_id is null (when
   responses aren't saved)
4. Conditional metadata storage: Added guard clause to prevent storing
   feedback email metadata when no post ID exists
5. Safe nonce generation: Added null check when generating nonces to
   avoid issues with missing post IDs
6. Action button handling: Made the "View in dashboard" action button
   conditional on having a valid dashboard URL

These changes ensure the email notification system works properly
regardless of whether form responses are being saved to the database or
not.

vianasw committed:
Fix forms failing tests

* Added $expected_attributes['saveResponses'] = 'yes'; to the expected
  attributes in the test.

vianasw committed:
Fix bad conflict resolution

I deleted too much during the rebase conflict resolution.

vianasw committed:
Add tests for saveResponses option in contact forms

- Add test_process_submission_stores_feedback_when_save_responses_yes()
  Tests that form submissions are stored when saveResponses='yes'
- Add
  test_process_submission_does_not_store_feedback_when_save_responses_no()
  Tests that form submissions are NOT stored when saveResponses='no'
- Add
  test_process_submission_stores_feedback_when_save_responses_default()
  Tests that form submissions are stored by default (saveResponses
  defaults to 'yes')

These tests ensure the saveResponses attribute properly controls whether
form submissions are stored in the WordPress database while maintaining
email functionality regardless of the setting.

simison commented on projects/packages/forms/src/blocks/contact-form/attributes.js:
```diff
@@ -52,4 +52,8 @@ export default {
 			listName: null,
 		},
 	},
+	saveResponses: {
+		type: 'string',
+		default: 'yes',
+	},
```
I understood there was a shortcodes-related legacy reason these had to be carried around as "yes/no" rather than `true/false`, but was it necessary for the block attributes as well? If in future we refactor away from shortcode-mangled way of doing this, we'll then need deprecation method for the attributes too instead of just refactoring the backend.

I see we use boolean in one of the attributes:

https://github.com/Automattic/jetpack/blob/dda00a325d883a530f3cd816d45990673200addb/projects/packages/forms/src/blocks/contact-form/attributes.js#L28-L31

Basically, is there a way to avoid having the attribute here as a string even if it is later carried around as a string? :-)

vianasw replied:
Good idea. I've addressed that in the latest commit.

vianasw committed:
Use new post status for temp to skip form inbox

- Move status determination logic from Feedback::save() to
  Contact_Form::process_submission()
- Centralize all status logic (trash, spam, temp, publish) in one
  location
- Remove form reference from Feedback class as it's no longer needed
- Improve code organization by keeping status logic with form processing
- Maintain existing functionality while improving maintainability

The temp status is now determined early in the form processing flow
and passed to the Feedback object, creating a cleaner separation
of concerns between form processing and data persistence.

vianasw committed:
Fix form submission handling saveResponses off

- Remove redundant $post_id check when updating unread feedback count
- Update dashboard URL condition to check feedback_status instead of
  $post_id
- Ensures proper handling of temporary feedback posts when form
  submissions are not stored # Please enter the commit message for your
  changes. Lines starting

vianasw committed:
Switch saveAttributes block attribute to boolean

In the backend we have to live with the 'yes'/'no' string because of the
shortcode legacy reason but we don't have to in the block attribute.
This way we'd have to just refactor the backend if we move away from the
shortcode-mangled way of doing things.

CGastrell commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -279,6 +279,21 @@ protected function __construct() {
 
+		// Add "temp" as a post status for temporary storage when saveResponses is 'no'
+		register_post_status(
+			'temp',
```
This worked fine, I just wonder if we should purposely use a less common word. After all, this will live both on dotcom and on any self hosted, we need to make sure the post status registration doesn't fail because there was already one there with the same name. I'd use some more complex, like jetpack-temp-feedback or something

vianasw replied:
hum, good point. I'll give it a try.

CGastrell reviewed (dismissed):
Worked well on all cases, including MailPoet integration even without saving the response. I left a comment for your consideration, but otherwise LGTM! :shipit:

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -279,6 +279,21 @@ protected function __construct() {
+				// translators: The temporary count.
+				'label_count'            => _n_noop( 'Temporary <span class="count">(%s)</span>', 'Temporary <span class="count">(%s)</span>', 'jetpack-forms' ),
+				'protected'              => true,
+				'_builtin'               => false,
```
I don't think we need to do this. Since this is the default for '_builtin' is false; 

Can we also set `internal` to true. So that the status work more as expected. 
It then future setting will be more align with what we expected. 

Such as `show_in_admin_status_list` - see https://developer.wordpress.org/reference/functions/register_post_status/

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -279,6 +279,21 @@ protected function __construct() {
+				'exclude_from_search'    => true,
+				'show_in_admin_all_list' => false,
+				// translators: The temporary count.
+				'label_count'            => _n_noop( 'Temporary <span class="count">(%s)</span>', 'Temporary <span class="count">(%s)</span>', 'jetpack-forms' ),
```
I don't think we need this. Not sure where it would show up. Since we should be hiding it.

vianasw replied:
You're right, we don't need it.

enejb commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -2076,15 +2099,14 @@ public function process_submission() {
 			'contact-form-sent' => $post_id,
 			'contact-form-hash' => $this->hash,
-			'_wpnonce'          => wp_create_nonce( "contact-form-sent-{$post_id}" ), // wp_nonce_url HTMLencodes :( .
+			'_wpnonce'          => $post_id ? wp_create_nonce( "contact-form-sent-{$post_id}" ) : '', // wp_nonce_url HTMLencodes :( .
```
Do we still need this. Since we should be always setting the $post_id now.

vianasw replied:
Yes, this is a leftover from my previous changes that we don't need anymore.

enejb commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1806,6 +1815,8 @@ public function process_submission() {
 		} elseif ( $is_spam ) {
 			$feedback_status = 'spam';
+		} elseif ( 'no' === $this->get_attribute( 'saveResponses' ) ) {
+			$feedback_status = 'temp';
```
Should we always be setting to 'temp' status? 

My assumtion would be that this would always happen if saveResponses is set to "no"

So even if we mark the post as "spam" we. Not sure if we need another status here. Such as "spam-temp" or something like that 🤔. I would assume no.

simison replied:
It's fine for spam to appear in Spam folder and then normal spam workflows can apply.

The question gets interesting when they mark it as "not spam"; in normal scenario we place the message back to inbox and send the email notification.

simison replied:
Spam flows are all fine to iterate in follow-up PRs though.

vianasw replied:
Good catch. The easiest thing would be to do as Enej suggests and skip the spam folder if saveResponses is set to 'no'. Do you foresee any side effects with that approach?

vianasw replied:
Ok, replying myself: one side effect of skipping the `spam` status and applying `temp` always would be that integrations won't benefit from the spam detection.

simison replied:
Yeah, this feature would go missing and with it ability to recover from false positives:

<img width="299" height="297" alt="image" src="https://github.com/user-attachments/assets/f5865162-361f-4766-a8d6-1a4fe4efc414" />

enejb commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1861,15 +1872,17 @@ public function process_submission() {
 		remove_filter( 'wp_insert_post_data', array( $plugin, 'insert_feedback_filter' ), 10 );
 
-		update_post_meta( $post_id, '_feedback_extra_fields', $this->addslashes_deep( $extra_values ) );
+		if ( $post_id ) {
```
DO we still need this check?

vianasw replied:
Well, the Feedback's save method could fail, right? In which case $post_id would be 0 and I'm not sure what would happen to update_post_meta with 0 as the post_id. It's definitely something that we didn't protect against but I'm unsure if it was just an issue we didn't know we had. I'll check what happens in that case.

vianasw replied:
I looked into this and the check is unnecessary. `update_post_meta` will check if the post exists and just return false and Post IDs are always positive integers starting from 1.

enejb commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1861,15 +1872,17 @@ public function process_submission() {
 		}
 
-		if ( defined( 'AKISMET_VERSION' ) ) {
+		if ( defined( 'AKISMET_VERSION' ) && $post_id ) {
```
The same as this?

vianasw replied:
I'm unsure, but I'll dig deeper.

vianasw committed:
Remove unnecessary label_count post status arg

vianasw committed:
Remove unnecessary defensive programming

vianasw committed:
Rename temp post status to a less generic name

We're using jetpack-temp-feedback to avoid clashing with other plugin
implementations out there that might have used the same generic name as
a post status.

vianasw committed:
Remove unnecessary post_id null checks

vianasw committed:
Rename temp feedback post status

I've renamed the post status we're using to skip the feedback inbox from
`jetpack-temp-feedback` to `jp-temp-feedback` because we were hitting
the character limit for post status (20 characters) and the posts
weren't saved because of that. Unit tests didn't pick this issue because
we use WorDBless to mock post insertion and this logic is missing.

enejb reviewed (approved):
Nice work! This works as expected for me.
